### DISCUSSION TITLE
Scaling of coloring step in `casadi.nlpsol`

### DISCUSSION THREAD

**Discussion by gregreich** [casadi/casadi#3424] [Q&A]
Hi

I have a non-linear constrained optimisation problem that I solve with `casadi.nlpsol`; to make the computations of realistically-sized experiments feasible, I had to optimise my implementations of both the objective function and the constraints using `casadi.Function.map` and `fold`. After those optimisations, memory usage appears to scale close to linearly; but the overall runtime of a problem—including the construction of the solver object—explodes when I scale the problem (a spline collocation problem with a maximum likelihood estimation wrapped around it).

When I look at the log, I realise that most of the time is spend in coloring (I hope I got all the relevant parts of it):
```diff
  CasADi - 2023-10-23 16:42:43 MESSAGE("Block size: 30") [.../casadi/core/function_internal.cpp:1267]
  CasADi - 2023-10-23 16:42:43 MESSAGE("Star coloring on 64x64: 63 <-> 64") [.../casadi/core/function_internal.cpp:1280]
  CasADi - 2023-10-23 16:42:50 MESSAGE("Number of sweeps: 33") [.../casadi/core/function_internal.cpp:1450]
  CasADi - 2023-10-23 16:42:50 MESSAGE("Formed Jacobian sparsity pattern (dimension [1896,1896], 1950264 (54.2521 %) nonzeros.") [.../casadi/core/function_internal.cpp:1452]
  CasADi - 2023-10-23 16:42:50 MESSAGE("Allocated return value") [.../casadi/core/x_function.hpp:464]
  CasADi - 2023-10-23 16:42:50 MESSAGE("helper_jacobian_MX::get_partition") [.../casadi/core/function_internal.cpp:1907]
! CasADi - 2023-10-23 16:42:50 MESSAGE("FunctionInternal::getPartition star_coloring") [.../casadi/core/function_internal.cpp:1920]
  CasADi - 2023-10-23 16:54:56 MESSAGE("Star coloring completed: 1392 directional derivatives needed (1896 without coloring).") [.../casadi/core/function_internal.cpp:1925]
  CasADi - 2023-10-23 16:54:56 MESSAGE("Graph coloring completed") [.../casadi/core/x_function.hpp:475]
```
Of the total wall-time of about 950 secs spend to solve this (reduced) problem, more than 720 secs go into coloring; some 210 secs go into the actual call to the solver (using a `threadMap`, so some parallelism involved). When I scale the problem up, I have a scaling worse than cubic, which quickly renders computation infeasible.

Does anybody have ideas or resources I can look up to address this issues? Any input greatly appreciated.

Thanks a lot!

**Comment by gregreich** [comment]
Update: This is the sparsity pattern of my Jacobian:
![sparsity](https://github.com/casadi/casadi/assets/23191215/aa437472-e9ae-40c9-8d99-6e77d9633c1e)
(Note that the numbers in the log output above and those from the plot do not exactly coincide, since they are obtained from different configurations of the problem.)

**Reply by jgillis** [reply]
The highlighted message is about the Hessian (computed as symmetric Jacobian of the gradient). Do you have a sparsity plot for that, too?
The star coloring implemented in CasADi is not particularly optimized, and not at all tailored for block dense systems.
I think you just hit a novel direction in which we don't currently scale well.

There is an old branch https://github.com/casadi/casadi/tree/colpack2 were coloring is performed as:
 * save sparsity to the disk
 * run Colpack
 * load back coloring from the disk

You could do as follows:
 * merge these commits to dvelop casadi
 * try out different colpack coloring routines
 * make the the colpack call cached (if a certain sparsity has been colored before, just use that older coloring)
 * best result is probably if you color yourself (looks trivial for your structure?)
 
A nice addition to casadi would be:
 * before coloring a sparsity pattern A, deflate the sparsity pattern: find Ad, m qand n such that `A = kron(Ad,ones(m,n))`
 * perform the coloring on Ad
 * inflate the result

Update: your latest pattern does not look trivial anymore

**Reply by gregreich** [reply]
Thank you very much for your prompt reply, Joris!

> The highlighted message is about the Hessian (computed as symmetric Jacobian of the gradient).

Thanks for the hint; that explains a lot of my confusion when making sense of the actual numbers in the log...

> Update: your latest pattern does not look trivial anymore

Sorry for the confusion; I hadn't looked at the pattern of my problem for a while, and so the previous version made me aware of some optimization to harvest more sparsity, which I forgot to apply in the that version. Having done that yields the current version, which is what is relevant for this discussion.

> I think you just hit a novel direction in which we don't currently scale well.

Maybe I did; but I might as well disappoint you, in the sense that my *problem* actually does not scale well. However, since slightly simpler—and vastly more popular—versions of it might indeed be affected by what you pointed out, I take it as an excuse for my lengthy answer...

Below are my Jacobian and Hessian patterns, respectively (although it's not, think of it as a zoomed-in version of the pattern above, covering the first 1000 variables approximately):
![sparsity_jac](https://github.com/casadi/casadi/assets/23191215/327fb7ac-ec9d-46e9-8fb1-b3be7b2cb490)
![sparsity_hess](https://github.com/casadi/casadi/assets/23191215/0461a143-3625-4ed3-ba0b-505d72090acb)

Clearly, the Hessian has a structure related to that of the Jacobian. In fact, for the *simpler* version of the problem that I referred to above, if `A` is a logical matrix representing the sparsity pattern of the Jacobian, the Hessian has a pattern very similar to `A|A'` (`A` OR `A` transposed). However, in my version of the problem, the Hessian is essentially dense.

(A little bit of background, skip if not interested: I'm solving an MPEC problem, where the objective is a likelihood function, and the constraints represent the model being solved. In my case, the likelihood is an *integrated* likelihood, i.e., there is a state of the model that has no observations in the data, and which is thus integrated out in the likelihood (but not in the constraints). Even worse, the unobserved state is, in my case at hand, affecting *all* observations simultaneously. Thus, the remaining sparsity only comes from model states that have no observations in the data set at all (but still need to be solved for in the model). Consequently, as the data set gets larger, covering more and more states, the Hessian gets denser...)

So what are the implications for the coloring step: When I look at the remainder of the log, my understanding is that 
- there isn't actually a whole lot of unnecessary derivative computation avoided (only about one third)
  ```diff
    CasADi - 2023-10-24 09:41:16 MESSAGE("Formed Jacobian sparsity pattern (dimension [1896,1896], 1950264 (54.2521 %) nonzeros.") [.../casadi/core/function_internal.cpp:1452]
    CasADi - 2023-10-24 09:41:16 MESSAGE("Allocated return value") [.../casadi/core/x_function.hpp:464]
    CasADi - 2023-10-24 09:41:16 MESSAGE("helper_jacobian_MX::get_partition") [.../casadi/core/function_internal.cpp:1907]
    CasADi - 2023-10-24 09:41:16 MESSAGE("FunctionInternal::getPartition star_coloring") [.../casadi/core/function_internal.cpp:1920]
  + CasADi - 2023-10-24 09:53:18 MESSAGE("Star coloring completed: 1392 directional derivatives needed (1896 without coloring).") [.../casadi/core/function_internal.cpp:1925]
  ```
- the actual *computation* of the derivates does not take a lot of time (relative to the coloring step itself—same problem as above, so after more than 700 secs of coloring)
  ```
  CasADi - 2023-10-24 09:53:18 MESSAGE("22 sweeps needed for 1392 forward and 0 reverse directions") [.../casadi/core/x_function.hpp:535]
  CasADi - 2023-10-24 09:53:18 MESSAGE("0 %") [.../casadi/core/x_function.hpp:549]
  CasADi - 2023-10-24 09:53:18 MESSAGE("Calling 'ad_forward'") [.../casadi/core/x_function.hpp:640]
   ...
  CasADi - 2023-10-24 09:53:22 MESSAGE("90 %") [.../casadi/core/x_function.hpp:549]
  CasADi - 2023-10-24 09:53:22 MESSAGE("Calling 'ad_forward'") [.../casadi/core/x_function.hpp:640]
   ...
  CasADi - 2023-10-24 09:53:22 MESSAGE("Back from 'ad_forward'") [.../casadi/core/x_function.hpp:642]
  CasADi - 2023-10-24 09:53:22 MESSAGE("nlp_hess_l::init") [.../casadi/core/function_internal.cpp:543]
  ```

So the obvious question is: Can't I just avoid coloring by using a dense Hessian, and, if so, how can I best implement that in Casadi?

Some further notes on what I've tried before:
- Currently, I call the solver using an exact Hessian; from an optimization perspective, this works extremely well, usually yielding quadratic convergence.
- I did try (proper) BFGS, which appeared to work fine as well (convergence is superlinear, as expected), but which doesn't help me much since the complete Hessian seems to be evaluated once at the beginning, presumably to initialise BFGS. Maybe this option would be the "sweet spot" when working with a truly dense Hessian.
- I also tried limited-memory BGFS, but I couldn't get it to converge at all.

Thank you so much again, and sorry for the lengthy reply.

**Reply by gregreich** [reply]
I just double checked the use of limited-memory BFGS as implemented in both KNITRO and IPOPT, and I noted that even under this configuration casadi obtains the coloring and derives the Hessian; from the debug output, I don't think that it is ever actually *evaluated*, though. IMHO, this doesn't make much sense. Therefore, even if my specific problem appears esoteric, any proper case for limited-memory BFGS would require the ability to turn off the up-front Hessian computation.

So I think this discussion goes beyond my individual case. 

My suggestion is to add an option `hess_type` (or similar) to `casadi.nlpsol`, taking one of three values:
- `'sparse'` (default): status quo
- `'dense'`: omit the coloring step, but nevertheless assemble and provide the Hessian when constructing the solver function object
- `identity`: neither color nor assemble the Hessian, but rather return an identity matrix of appropriate size. Having this option would ensure compatibility with solvers that call the Hessian as 'Step 0' for their BFGS approximation (e.g., to give the user a choice what to use). Many solvers use the identity matrix as an initial BFGS approximation of the Hessian by themselves anyway, but making this explicit safeguards from solver calls to a non-existing object—without having casadi to interpret any solver-specific option regarding its use of the Hessian.

Does that sound reasonable, @jgillis?

**Comment by jaeandersson** [comment]
> I just double checked the use of limited-memory BFGS as implemented in both KNITRO and IPOPT, and I noted that even under this configuration casadi obtains the coloring and derives the Hessian; from the debug output, I don't think that it is ever actually _evaluated_, though. IMHO, this doesn't make much sense. Therefore, even if my specific problem appears esoteric, any proper case for limited-memory BFGS would require the ability to turn off the up-front Hessian computation.
> 
> So I think this discussion goes beyond my individual case.
> 
> My suggestion is to add an option `hess_type` (or similar) to `casadi.nlpsol`, taking one of three values:
> 
> * `'sparse'` (default): status quo
> * `'dense'`: omit the coloring step, but nevertheless assemble and provide the Hessian when constructing the solver function object
> * `identity`: neither color nor assemble the Hessian, but rather return an identity matrix of appropriate size. Having this option would ensure compatibility with solvers that call the Hessian as 'Step 0' for their BFGS approximation (e.g., to give the user a choice what to use). Many solvers use the identity matrix as an initial BFGS approximation of the Hessian by themselves anyway, but making this explicit safeguards from solver calls to a non-existing object—without having casadi to interpret any solver-specific option regarding its use of the Hessian.
> 
> Does that sound reasonable, @jgillis?

I think you're making it way too complicated. All you need is a simple option to disable the graph coloring, which could be added in like 1h. The `identity` option above is of little use: You can't provide the Jacobian routines with a less sparse pattern - it will cause errors in the calculations. If you wanted to provide an identity Hessian, you can already provide it by defining a custom Hessian function.

Anyway, you can also work around this whole issue by providing your own Jacobian/Hessian routine with existing code.

**Reply by gregreich** [reply]
Thanks, @jaeandersson, for your reply.

>  The `identity` option above is of little use: You can't provide the Jacobian routines with a less sparse pattern

Maybe there is a misunderstanding here: I'm only talking about Hessians (i.e., the Jacobian of the gradient of either the objective function alone, or the Lagrangian—that's why I called it `hess_type`); but without such option, the use of *limited-memory* BFGS—which is an option many solvers (incl. KNITRO and IPOPT) provide—from within Casadi would still either be inefficient (because the derivative would still be constructed, even if coloring has been disabled), or more cumbersome than necessary, because one would have to provide a custom Hessian via `casadi.Function.cache` option just to get the identity matrix. Actually, this is what I do right now to get the problem stated above feasible, but at the expense of 6 additional lines of code—just to get an identity matrix in the right place...
